File: Baekjoon/3665_최종순위/B_3665.js
const isCycle = (n, degree) => {
  for (let i = 1; i < n; i++) {
    if (degree[i] > 0) {
      return true;
    }
  }

  return false;
}

const getStartNode = (n, degree) => {
  const startNode = [];

  for (let i = 1; i <= n; i++) {
    if (!degree[i]) {
      startNode.push(i);
    }
  }

  return startNode;
}

const topologySort = (n, degree, grade) => {
  const answer = [];

  const q = getStartNode(n, degree);
  
  while (q.length > 0) {
    const start = q.shift();

    answer.push(start.toString());

    for (let end = 1; end <= n; end++) {
      if (grade[start][end]) {
        degree[end] -= 1;

        if (!degree[end]) {
          q.push(end);
        }
      }
    }
  }

  if (isCycle(n, degree)) {
    return 'IMPOSSIBLE';
  }

  return answer.join(' ');
}

// const fs = require('fs');
// const input = fs.readFileSync('/dev/stdin').toString().trim().split('\n');

const input = [
  '3',
  '5',
  '5 4 3 2 1',
  '2',
  '2 4',
  '3 4',
  '3',
  '2 3 1',
  '0',
  '4',
  '1 2 3 4',
  '3',
  '1 2',
  '3 4',
  '2 3',
];

// const input = [
//   '1',
//   '3',
//   '2 3 1',
//   '0',
// ];

const t = parseInt(input[0]);

let idx = 1;

for (let i = 0; i < t; i++) {
  const n = parseInt(input[idx]);

  idx += 1;

  const inDegree = new Array(n + 1).fill(0);

  const grades = new Array(n + 1).fill(0).map(v => new Array(n + 1).fill(0));

  const lastGrades = input[idx].split(' ').map(v => parseInt(v));

  for (let i = 0; i < n - 1; i++) {
    const f = lastGrades[i];

    for (let j = i + 1; j < n; j++) {
      const g = lastGrades[j];

      inDegree[g] += 1;
      grades[f][g] = 1;
    }
  }

  idx += 1;

  const m = parseInt(input[idx]);

  idx += 1;

  const gradeChangedTeams = [];

  if (m > 0) {
    for (let j = idx; j < idx + m; j++) {
      const changedTeams = input[j].split(' ').map(v => parseInt(v))
  
      gradeChangedTeams.push(changedTeams);
  
      const [f, g] = changedTeams;
  
      if (grades[f][g]) {
        grades[f][g] = 0;
        grades[g][f] = 1;
  
        inDegree[g] -= 1;
        inDegree[f] += 1;
      } else {
        grades[f][g] = 1;
        grades[g][f] = 0;
  
        inDegree[g] += 1;
        inDegree[f] -= 1;
      }
    }
  
    idx += m;
  }

  console.log(topologySort(n, inDegree, grades));
}

// https://deque.tistory.com/62

// disjoint-set 을 생각했으나
// 이는 포함 관계를 정리하기 위한
// 부분 집합이 필요할때 사용하기에 적절하나
// 이번 문제처럼 우선 순위를 판단해야할 문제에는
// 위상정렬을 사용하는게 적절하다

File: baekjoon/3665_최종순위/B_3665.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        Main main = new Main();
        main.solution();
    }

    public static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

    private void solution() throws IOException {
        int t = Integer.parseInt(br.readLine());

        for (int i = 0; i < t; i++) {
            System.out.println(eachSolution());
        }
    }

    private String eachSolution() throws IOException {
        int n = Integer.parseInt(br.readLine());
        StringTokenizer rank = new StringTokenizer(br.readLine());

        int[] inDegree = new int[n + 1];
        ArrayList<Integer> tempRank = new ArrayList<>();
        int[][] priority = new int[n + 1][n + 1];

        for (int i = 0; i < n; i++) {
            int num = Integer.parseInt(rank.nextToken());
            inDegree[num] = i;
            tempRank.add(num);
        }

        for (int i = 0; i < tempRank.size() - 1; i++) {
            int from = tempRank.get(i);

            for (int j = i + 1; j < tempRank.size(); j++) {
                int to = tempRank.get(j);
                priority[from][to] = 1;
            }
        }

        int m = Integer.parseInt(br.readLine());

        for (int i = 0; i < m; i++) {
            StringTokenizer changedRank = new StringTokenizer(br.readLine());
            int team1 = Integer.parseInt(changedRank.nextToken());
            int team2 = Integer.parseInt(changedRank.nextToken());

//             inDegree 로 아래처럼 비교할 경우
//             inDegree[team1] == inDegree[team2] 인 경우를
//             제대로 고려하지 못하고 else 문으로 들어간다
//             5
//             5 4 3 2 1
//             2
//             2 4
//             3 4
//             해당 입력에서
//             차수 정보는 [0, 1, 2, 3, 4] 인데
//             2 4 에서 [0, 2, 2, 2, 4] 가 되고
//             3 4 에서 3 과 4 의 차수가 같다
//
//             inDegree 는 차수 정보만 갖고 있고
//             둘 사이의 상대적인 우선 순위는 제대로 파악할 수 없다
//             (위에서 본 것처럼 차수 정보가 같으면 어떤게 우선 순위가 높은지 알 수 없다)
//             inDegree 는 같은 경우도 있기 때문에
//             차수 정보로 우선 순위를 파악하지 않고
//             서로 간의 우선 순위가 항상 구분이 되도록
//             별도로 우선 순위를 판단해야 한다
//
//            if (inDegree[team1] > inDegree[team2]) {
//                inDegree[team1] -= 1;
//                inDegree[team2] += 1;
//            } else {
//                inDegree[team1] += 1;
//                inDegree[team2] -= 1;
//            }

            if (priority[team1][team2] == 1) {
                priority[team1][team2] = 0;
                priority[team2][team1] = 1;

                inDegree[team1] += 1;
                inDegree[team2] -= 1;
            } else {
                priority[team1][team2] = 1;
                priority[team2][team1] = 0;

                inDegree[team1] -= 1;
                inDegree[team2] += 1;
            }
        }

        ArrayList<Integer> list = topologySort(inDegree, priority);

        if (list.isEmpty()) {
            return "IMPOSSIBLE";
        }

        StringBuilder sb = new StringBuilder();

        for (Integer integer : list) {
            sb.append(integer);
            sb.append(" ");
        }

        return sb.toString();
    }

    private ArrayList<Integer> topologySort(int[] degree, int[][] priorityInfo) {
        int go = findStart(degree);

        if (go == -1) {
            return new ArrayList<>();
        }

        Deque<Integer> deq = new ArrayDeque<>();
        deq.add(go);
        ArrayList<Integer> answer = new ArrayList<>();

        while (!deq.isEmpty()) {
            int start = deq.pollFirst();

            answer.add(start);

            for (int end = 1; end < degree.length; end++) {
                if (priorityInfo[start][end] == 1) {
                    degree[end] -= 1;

                    if (degree[end] == 0) {
                        deq.add(end);
                    }
                }
            }
        }

        if (isCycle(degree)) {
            return new ArrayList<>();
        }

        return answer;
    }

    private int findStart(int[] degree) {
        for (int i = 1; i < degree.length; i++) {
            if (degree[i] == 0) {
                return i;
            }
        }

        return -1;
    }

    private boolean isCycle(int[] degree) {
        for (int i = 1; i < degree.length; i++) {
            if (degree[i] != 0) {
                return true;
            }
        }

        return false;
    }
}
